README.txt
SVGGraph Library version 2.15.1
===============================

This library provides PHP classes and functions for easily creating SVG
graphs from data. As of version 2.0, SVGGraph works with PHP 5 only -
PHP 4 support has been dropped.

Here is a basic example:
 $graph = new SVGGraph(640, 480);
 $graph->colours = array('red','green','blue');
 $graph->Values(100, 200, 150);
 $graph->Links('/Tom/', '/Dick/', '/Harry/');
 $graph->Render('BarGraph');

Full documentation is available at http://www.goat1000.com/

Graph types
===========
At the moment these types of graph are supported by SVGGraph:
 
 BarGraph        - vertical bars, optionally hyperlinked;

 LineGraph       - a line joining the data points, with optionally hyperlinked 
                   markers at the data points;

 PieGraph        - a pie chart, with optionally hyperlinked slices and option
                   to fade labels in/out when the pointer enters/leaves a
                   slice;

 Bar3DGraph      - a 3D-looking version of the BarGraph type;

 Pie3DGraph      - a 3D-looking version of the PieGraph type;

 ScatterGraph    - markers drawn at arbitrary horizontal and vertical points;

 MultiLineGraph  - multiple data sets drawn as lines on one graph;

 StackedBarGraph - multiple data sets drawn as bars, stacked one on top of
                   another;

 GroupedBarGraph - multiple data sets drawn as bars, side-by-side;

 StackedLineGraph - multiple data sets, their values added together;

 MultiScatterGraph - scatter graph supporting multiple data sets;

 HorizontalBarGraph - a bar graph with the axes swapped;

 HorizontalStackedBarGraph - a stacked bar graph drawn horizontally;

 HorizontalGroupedBarGraph - a grouped bar graph drawn horizontally;

 RadarGraph - a radar or star graph with values drawn as lines;

 MultiRadarGraph - a radar graph supporting multiple data sets;

 CylinderGraph - a 3D bar graph with the bars cylinder shaped;

 StackedBar3DGraph - a 3D bar graph version of the stacked bar graph;

 GroupedBar3DGraph - a 3D bar graph version of the grouped bar graph;

 StackedCylinderGraph - a cylinder-bar version of the stacked bar graph;

 GroupedCylinderGraph - a cylinder-bar version of the grouped bar graph;

 DonutGraph - a pie graph with a hole in the middle;

 PolarAreaGraph - a pie graph where the area of the slice varies instead of
                  its angle;

 ExplodedPieGraph - a pie graph with slices exploded out from the centre.
 
Using SVGGraph
==============
The library consists of several class files which must be present. To use
SVGGraph, include or require the SVGGraph.php class file. The other classes
should be in the same directory as this main file to be loaded automatically.

Embedding SVG in a page
=======================
There are several ways to insert SVG graphics into a page. FireFox, Safari,
Chrome, Opera all support SVG, though Internet Explorer currently requires
the use of a plugin (supplied by Adobe).

For options 1-3, I'll assume you have a PHP script called "graph.php" which
contains the code to generate the SVG.

Option 1: the embed tag
 <embed src="graph.php" type="image/svg+xml" width="600" height="400"
  pluginspage="http://www.adobe.com/svg/viewer/install/" />

This method works in all browsers, though the embed tag is not part of the HTML
standard.

Option 2: the iframe tag
 <iframe src="graph.php" type="image/svg+xml" width="600" height="400"></iframe>

This method also works in all browsers, and the iframe tag is standard.

Option 3: the object tag
 <object data="graph.php" width="600" height="100" type="image/svg+xml" />

The object tag is standard, but this doesn't work in IE.

Option 4: using the svg namespace within an xhtml document

This option is more complicated, as it requires changing the doctype and
content type of the page being served. The SVG is generated as part of the
same page.
 <?php
  header('content-type: application/xhtml+xml; charset=UTF-8');
  // $graph = new SVGGraph(...);
  // $graph setup here!
 ?>
 <!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1 plus MathML 2.0 plus SVG 1.1//EN"
  "http://www.w3.org/2002/04/xhtml-math-svg/xhtml-math-svg.dtd">
 <html xmlns="http://www.w3.org/1999/xhtml"
  xmlns:svg="http://www.w3.org/2000/svg"
  xmlns:xlink="http://www.w3.org/1999/xlink" xml:lang="en">
 <head>
  <meta http-equiv="Content-Type" content="application/xhtml+xml; charset=UTF-8" />
  <title>SVGGraph example</title>
 </head>
 <body>
  <h1>Example of SVG in XHTML</h1>
  <div>
  <?php echo $graph->Fetch('BarGraph', false); ?>
  </div>
 </body>
 </html>

This method allows you more control over how you use the SVG, though again it
doesn't work in IE.

Class Constructor
=================
The SVGGraph class constructor takes three arguments, the width and height 
of the SVG image in pixels and an optional array of settings to be passed to 
the rendering class.
 $graph = new SVGGraph($width, $height, $settings);

For more information on the $settings array, see the section below.

Data Values
===========
For simple graphs you may set the data to use by passing it into the Values
function:
 $graph->Values(1, 2, 3);

For more control over the data, and to assign labels, pass the values in as an
array:
 $data = array('first' => 1, 'second' => 2, 'third' => 3);
 $graph->Values($data);

For graphs supporting multiple datasets, pass each dataset as an array within
an outer array:
 $data = array(
  array('first' => 1, 'second' => 2, 'third' => 3),
  array('first' => 3, 'second' => 4, 'third' => 2)
 );
 $graph->Values($data);

Scatter graphs draw markers at x,y coordinates, given as the key and value in
the data array:
 $data = array(5 => 20, 6 => 30, 10 => 90, 20 => 50);
 $graph->Values($data);

This will draw the markers at (5,20), (6,30), (10,90) and (20,50). The new 
scatter_2d option in version 2.4 allows points to occupy the same x-
coordinate by passing the values as pairs of coordinates:
 $graph = new SVGGraph(200,100, array('scatter_2d' => true));
 $data = array(array(5,20), array(6,30), array(5,90), array(10,50));
 $graph->Values($data);

Note: data in this format are not supported by any of the non-scatter graph
types.

Hyperlinks
==========
The graph bars and markers may be assigned hyperlinks - each value that requires
a link should have a URL assigned to it using the Links function:
 $graph->Links('/page1.html', NULL, '/page3.html');

The NULL is used here to specify that the second bar will not be linked to
anywhere.

As with the Values function, the list of links may be passed in as an array:
 $links = array('/page1.html', NULL, '/page3.html');

Using an associative array means that NULL values may be skipped.
 $links = array('first' => '/page1.html', 'third' => '/page3.html');
 $graphs->Links($links);

Rendering
=========
To generate and display the graph, call the Render function passing in the
type of graph to be rendered:
 $graph->Render('BarGraph');

This will send the correct content type header to the browser and output the
SVG graph.

The Render function takes two optional parameters in addition to the graph
type:
 $graph->Render($type, $header, $content_type);

Passing in FALSE for $header will prevent output of the XML declaration and
doctype. Passing in FALSE for $content_type will prevent the 'image/svg+xml'
content type being set in the response header.

To generate the graph without outputting it to the browser you may use the
Fetch function instead:
 $output = $graph->Fetch('BarGraph');

This function also takes an optional $header parameter:
 $output = $graph->Fetch($type, $header);

Passing in FALSE as $header will prevent the returned output from containing
the XML declaration and doctype. The Fetch function never outputs the content
type to the response header.

Colours
=======
The colours used may be overridden from the default random set by setting the
graph's "colours" array.
 $graph->colours = array('red', 'green', '#00ffff', 'rgb(100,200,100)',
    array('red','green'));

You may use any of the standard named colours, or hex notation, or RGB notation.

Gradients
=========
The final entry in the example colours array above is an array of two colours,
which specifies a vertical gradient, the first colour (red) at the top and the
second (green) at the bottom. From version 2.1, more colours may be used and
a final 'h' or 'v' will specify horizontal or vertical gradients.

Gradients are supported by the bar graphs and for the filled area under line
graphs. Where gradients are not supported, the first colour in the array will
be used instead.

Examples:
 array('red','white','red','h');
 - a horizontal gradient, red at both sides and white in the centre

 array('red','white','blue');
 - a vertical gradient, red at the top, white in the centre, blue at the bottom

 array('red','orange','yellow','green','blue','indigo','violet','h');
 - a horizontal rainbow gradient

Settings
========
Many of the ways that things are displayed may be changed by passing in an array
of settings to the SVGGraph constructor:
 $settings = array('back_colour' => 'white');
 $graph = new Graph($w, $h, $settings);

There are now too many options for me to list them all here without wasting a 
lot of time, and probably making mistakes. For the full list of options, examples
and descriptions, please visit the website: http://www.goat1000.com/svggraph.php


Contact details
===============
For more information about this software please contact the author,
graham(at)goat1000.com or visit the website: http://www.goat1000.com/


Copyright (C) 2009-2014 Graham Breach


CHANGES.txt
Version 2.15.1 (12/05/2014)
--------------
- Added crosshair options.
- Fixed X axis being incorrectly labelled when using large associative dataset.

Version 2.15 (11/03/2014)
------------
- Added donut graph.
- Added polar area graph.
- Added exploded pie graph.
- Added support for dual Y axes.
- Fixed legend dragging when embedded in HTML.

Version 2.14 (05/10/2013)
------------
- Added pattern fill support.
- Added total labels for stacked bar graphs.
- Better validation of structured data.
- Better handling of non-ASCII strings.
- Replaced rgba() values with opacity attributes.
- Fixed compatibility with PHP 5.1.x.
- Fixed drawing of pie graphs containing relatively small values.
- Fixed X axis calculations for negative or non-integral values.
- Fixed structured axis_text when units_label or units_before_label used.
- Fixed position of labels on stacked bar graphs with negative values.
- Fixed legend entries skipping NULL or 0 values.

Version 2.13 (31/07/2013)
------------
- Added PopulationPyramid graph.
- Added units_before_x and units_before_y options for axis text.
- Added units_label and units_before_label options for pie/bar labels.
- Added units_tooltip, units_before_tooltip, units_tooltip_key and
  units_before_tooltip_key options for tooltip values.
- Added stroke_dash, minify and minify_js options.
- Added structured data options: marker_type, marker_size, marker_stroke_width,
  marker_stroke_colour, stroke_colour, stroke_width and stroke_dash.
- Fixed display of character entities in tooltips.
- Fixed measurement of text containing character entities.
- Fixed dragging of legend over pie graphs.

Version 2.12 (18/05/2013)
------------
- Added log_axis_y and log_axis_y_base options.
- Added decimal and thousands options for number formatting.
- Added grid_back_stripe and grid_back_stripe_colour options.
- Added per-axis font options.
- Added units_x and units_y options for axis text.
- Added show_label_key option for pie graphs.
- 3D pie graphs now calculate a depth when the setting is too great.
- Fixed structure option to take precedence over scatter_2d option.
- Fixed display of small numbers on axes.
- Added workaround for Safari 6 <use/> bug.

Version 2.11 (12/03/2013)
------------
- Added stacked and grouped cylinder graphs.
- Added structured data support.
- Added normal and horizontal floating bar graphs.
- Added bubble graph.
- Added box and whisker graph.
- Added back_round_clip option.
- Added reverse option for radar graphs.
- Added support for negative values on radar graphs.
- Added support for multi-line tooltips.
- Fixed drawing of axes when all values negative.
- Fixed svg node in namespaced XHTML.
- Improved performance.

Version 2.10 (23/10/2012)
------------
- Added cylinder graph.
- Added stacked and grouped 3D bar graphs.
- Added support for gradients in graph element backgrounds.
- Added gradient shaded side to 3D pie graphs.
- Added best fit lines for scatter graphs.
- Added support for drawing associative data on scatter graphs.
- Added show_axis_h and show_axis_v options.
- Added minimum_units_y option.
- Improved axis text label spacing again.
- Better text positioning for radar graphs.
- New default colours.
- Fixed tooltips positioning when embedded in HTML5/XHTML.
- Fixed text being selected when dragging legend.
- Fixed graphs with all values negative not being drawn.

Version 2.9 (29/08/2012)
-----------
- Added bar label options.
- Added axis division and subdivision styles and other options.
- Added diamond, hexagon, octagon, asterisk, star, threestar, fourstar and
  eightstar markers.
- Added legend columns option.
- Added guide line opacity, text opacity, font adjust, text align, length and
  length in units options.
- Added grid background colour option.
- Added grid line style options.
- Added marker border options.
- Added character encoding option.
- Added gradient stop opacity support.
- Improved subdivision calculations.
- Improved axis text label spacing.
- Removed main clipping path when not using a rounded rectangle background.
- Fixed grid clipping path using non-unique ID.
- Fixed line graphs not reaching right hand side of graph.
- Fixed line graphs drawing vertical line to axis when filling enabled.
- Fixed guide line title line spacing.
- Fixed error caused by empty data sets.

Version 2.8 (27/07/2012)
-----------
- Added stacked line graph.
- Added radar graph and multi-radar graph.
- Added axis_text_space option.
- Added axis_stroke_width option.
- Added force_assoc option.
- Fixed legend dragging without tooltips enabled.
- Fixed display of labels when axes disabled.
- Fixed use of associative array data.
- Added fill to legend display for line graphs.

Version 2.7.1 (02/07/2012)
-------------
- Fixed order of entries in multi-line graphs.
- Fixed script type attribute to match SVG standard.
- Fixed order of legend entries for stacked bar, horizontal bar and
  horizontal grouped bar graphs.
- Fixed format of negative decimal numbers.

Version 2.7 (25/05/2012)
----------
- Added guideline options.
- Added axis text rotation.
- Improved HTML embedding.
- Reduced output size of grouped and stacked bar graphs.

Version 2.6 (30/03/2012)
-----------
- Moved all configuration defaults to svggraph.ini.
- Added legend_* options.
- Added label_* options.
- Added graph_title_* options.

Version 2.5.1 (20/02/2012)
-------------
- Fixed bug with 0 values in stacked graphs.

Version 2.5 (09/08/2011)
-----------
- Improved Javascript event handlers.
- Added grid and axis subdivision options.
- Added stroke width option.
- Added more line/scatter marker symbols.
- Added line dash option.
- Added support for per-dataset fill opacity and line width.
- Added pie graph start angle option.

Version 2.4 (07/07/2011)
-----------
- Added multiple scatter graph.
- Added option to use full x,y data for scatter graphs.
- Added support for per-dataset marker options.
- Added support for per-dataset fill_under options.

examples/scaling.php
<?php

require_once(__DIR__ . '/../SVGGraph.php');

$settings = array(
  'back_colour' => '#fff',  'stroke_colour' => '#000',
  'back_stroke_width' => 0, 'back_stroke_colour' => '#eee',
  'axis_colour' => '#333',  'axis_overlap' => 2,
  'axis_font' => 'Georgia', 'axis_font_size' => 10,
  'grid_colour' => '#ddd',  'label_colour' => '#000',
  'pad_right' => 20,        'pad_left' => 20,
  'link_base' => '/',       'link_target' => '_top',
  'minimum_grid_spacing' => 20,

  'crosshairs' => true,

  'preserve_aspect_ratio' => 'xMidYMid meet',
  'auto_fit' => true,
);

$settings['legend_entries'] = array('First series', 'Second series', 'Third series');

$values = array(
  array('Dough' => 30, 'Ray' => 50, 'Me' => 40, 'So' => 25, 'Far' => 45, 'Lard' => 35),
  array('Dough' => 20, 'Ray' => 30, 'Me' => 20, 'So' => 15, 'Far' => 25, 'Lard' => 35, 'Tea' => 45),
  array('Dough' => 11, 'Ray' => 11, 'Me' => 20, 'So' => 7, 'Far' => 40, 'Lard' => 22, 'Tea' => 1),
);

// Colors are copied from D3 that used http://colorbrewer2.org by Cynthia Brewer, Mark Harrower and The Pennsylvania State University
$colours = array(
  '#1f77b4', '#ff7f0e', '#2ca02c', '#d62728', '#9467bd',
  '#8c564b', '#e377c2', '#7f7f7f', '#bcbd22', '#17becf');

$links = array('Dough' => 'jpegsaver.php', 'Ray' => 'crcdropper.php',
               'Me' => 'svggraph.php');

if (!empty($_GET['object'])) {
  $graph = new SVGGraph(1000, 300, $settings);
  $graph->colours = $colours;
  $graph->Values($values);
  $graph->Links($links);
  $graph->Render('GroupedBarGraph');
  die;
}

/*

Some more reading:
 * http://thatemil.com/blog/2014/04/06/intrinsic-sizing-of-svg-in-responsive-web-design/
 * http://alistapart.com/article/creating-intrinsic-ratios-for-video
 * http://stackoverflow.com/questions/25566109/chrome-bug-in-svg-getscreenctm

*/

?>
<!DOCTYPE html>
<html  dir="ltr" lang="en" xml:lang="en">
<head>
  <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
  <style>
    .parentwithheight {
      height: 100px;
    }
    .parentwithwidth {
      width: 500px;
    }
    @media screen and (-ms-high-contrast: active), (-ms-high-contrast: none) {
      /* IE10+ specific styles go here */
      .wrapperdiv {
        position: relative;
        padding-bottom: 30%;  /* This is the ration of original svg size 1000x300 */
        height: 0;
      }
      .wrapperdiv svg,
      .wrapperdiv object {
        position: absolute;
        top: 0;
        left: 0;
        width: 100%;
        height: 100%;
      }
    }
  </style>
  <!--[if IE]>
  <style>
    /* IE < 10 styles */
    .wrapperdiv {
      position: relative;
      padding-bottom: 30%;  /* This is the ration of original svg size 1000x300 */
      height: 0;
    }
    .wrapperdiv svg,
    .wrapperdiv object {
      position: absolute;
      top: 0;
      left: 0;
      width: 100%;
      height: 100%;
    }
  </style>
  <![endif]-->
</head>
<body>
<h1>Scaling sample</h1>
<h2>Scaled to parent with height 100</h2>
<h3>Inline SVG</h3>
<div class="parentwithheight">
  <?php
  $graph = new SVGGraph(1000, 300, $settings);
  $graph->colours = $colours;
  $graph->Values($values);
  $graph->Links($links);
  echo $graph->Fetch('GroupedBarGraph', false, false);
  ?>
</div>
<h3>SVG embedded as object</h3>
<div class="parentwithheight">
  <object type="image/svg+xml" data="scaling.php?object=1" width="100%" height="100%">svg object</object>
</div>

<h2>Scaled to parent with width 500</h2>
<h3>Inline SVG</h3>
<div class="parentwithwidth">
  <?php
  $graph = new SVGGraph(1000, 300, $settings);
  $graph->colours = $colours;
  $graph->Values($values);
  $graph->Links($links);
  echo $graph->Fetch('GroupedBarGraph', false, false);
  ?>
</div>
<h3>SVG embedded as object</h3>
<div class="parentwithwidth">
  <object type="image/svg+xml" data="scaling.php?object=1" width="100%" height="100%">svg object</object>
</div>

<h2>Scaled to page width</h2>
<h3>Inline SVG</h3>
<div class="wrapperdiv">
  <?php
  $graph = new SVGGraph(1000, 300, $settings);
  $graph->colours = $colours;
  $graph->Values($values);
  $graph->Links($links);
  echo $graph->Fetch('GroupedBarGraph', false, false);
  ?>
</div>
<h3>SVG embedded as object</h3>
<div class="wrapperdiv">
  <object type="image/svg+xml" data="scaling.php?object=1" width="100%" height="100%">svg object</object>
</div>
<h2>Pie Graph</h2>
  <div class="wrapperdiv">
    <?php
    $settings = array();
    $settings['label_fade_in_speed'] = 30;
    $settings['label_fade_out_speed'] = 15;
    $settings['preserve_aspect_ratio'] = 'xMidYMid meet';
    $settings['auto_fit_parent'] = true;
    $settings['units_before_tooltip'] = '&#xa3;';

    $graph = new SVGGraph(1000, 300, $settings);
    $graph->colours = $colours;
    $graph->Values($values[0]);
    echo $graph->Fetch('PieGraph', false, false);
    ?>
  </div>
</body>
</html>
